python/ray/worker.py

```diff
@@ -1242,6 +1251,36 @@ def connect(node,
     ray.state.state._initialize_global_state(
         node.redis_address, redis_password=node.redis_password)
 
+    pkg_file = get_project_package_name(job_id.hex() if mode == SCRIPT_MODE
+                                        else os.getenv("RAY_JOB_ID"))
+    pkg_file_path = str(pkg_file)
+    if (required_directories or required_modules) and mode == SCRIPT_MODE:
+        # Create package if defined and ship it to remote storage
+        create_project_package(pkg_file, required_directories,
+                               required_modules)
+        # Push the data to remote storage
+        data = pkg_file.read_bytes()
+        internal_kv._internal_kv_put(pkg_file_path, data)
+        logger.debug(
+            f"Ray has packaged {required_directories} and {required_modules} "
+            "into {pkg_file_path} in {len(data)} bytes")
+    elif mode in (WORKER_MODE, RESTORE_WORKER_MODE, SPILL_WORKER_MODE):
+        internal_kv._internal_kv_get(pkg_file_path)
+        # For each node, the package will only be downloaded one time
+        # Locking to avoid multiple process download concurrently
+        lock = FileLock(pkg_file_path + ".lock")
+        with lock:
```

Comment:
Should we acquire the lock before doing the get, to avoid redundant data transfers?

Reply:
Good point!